controllers/operatorconfigmap_controller.go

```diff
@@ -123,204 +124,196 @@ func (c *ClusterVersionReconciler) SetupWithManager(mgr ctrl.Manager) error {
+			return ctrl.Result{}, err
+		}
 
-	rbdCSIDriver := templates.RbdCSIDriver.DeepCopy()
-	rbdCSIDriver.ObjectMeta.Name = csi.GetRBDDriverName()
-	err = csi.CreateCSIDriver(c.ctx, c.Client, rbdCSIDriver)
-	if err != nil {
-		c.log.Error(err, "unable to create rbd CSIDriver")
-		return ctrl.Result{}, err
-	}
+		// Need to handle deletion of the csiDriver object, we cannot set
+		// ownerReference on it as its cluster scoped resource
+		cephfsCSIDriver := templates.CephFSCSIDriver.DeepCopy()
+		cephfsCSIDriver.ObjectMeta.Name = csi.GetCephFSDriverName()
+		if err := csi.CreateCSIDriver(c.ctx, c.Client, cephfsCSIDriver); err != nil {
+			c.log.Error(err, "unable to create cephfs CSIDriver")
+			return ctrl.Result{}, err
+		}
 
-	prometheusRule := &monitoringv1.PrometheusRule{}
-	err = k8sYAML.NewYAMLOrJSONDecoder(bytes.NewBufferString(string(pvcPrometheusRules)), 1000).Decode(prometheusRule)
-	if err != nil {
-		c.log.Error(err, "Unable to retrieve prometheus rules.", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
-		return ctrl.Result{}, err
-	}
+		rbdCSIDriver := templates.RbdCSIDriver.DeepCopy()
+		rbdCSIDriver.ObjectMeta.Name = csi.GetRBDDriverName()
+		if err := csi.CreateCSIDriver(c.ctx, c.Client, rbdCSIDriver); err != nil {
+			c.log.Error(err, "unable to create rbd CSIDriver")
+			return ctrl.Result{}, err
+		}
 
-	operatorConfig, err := c.getOperatorConfig()
-	if err != nil {
-		return ctrl.Result{}, err
-	}
-	prometheusRule.SetNamespace(c.OperatorNamespace)
+		prometheusRule := &monitoringv1.PrometheusRule{}
+		if err := k8sYAML.NewYAMLOrJSONDecoder(bytes.NewBufferString(string(pvcPrometheusRules)), 1000).Decode(prometheusRule); err != nil {
+			c.log.Error(err, "Unable to retrieve prometheus rules.", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+			return ctrl.Result{}, err
+		}
 
-	err = c.createOrUpdate(prometheusRule, func() error {
-		applyLabels(operatorConfig.Data["OCS_METRICS_LABELS"], &prometheusRule.ObjectMeta)
-		return c.own(prometheusRule)
-	})
-	if err != nil {
-		c.log.Error(err, "failed to create/update prometheus rules")
-		return ctrl.Result{}, err
-	}
+		prometheusRule.SetNamespace(c.OperatorNamespace)
 
-	c.log.Info("prometheus rules deployed", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+		err = c.createOrUpdate(prometheusRule, func() error {
+			applyLabels(c.OperatorConfigMap.Data["OCS_METRICS_LABELS"], &prometheusRule.ObjectMeta)
+			return c.own(prometheusRule)
+		})
+		if err != nil {
+			c.log.Error(err, "failed to create/update prometheus rules")
+			return ctrl.Result{}, err
+		}
 
+		c.log.Info("prometheus rules deployed", "prometheusRule", klog.KRef(prometheusRule.Namespace, prometheusRule.Name))
+	}
 	return ctrl.Result{}, nil
 }
 
-func (c *ClusterVersionReconciler) createOrUpdate(obj client.Object, f controllerutil.MutateFn) error {
+func (c *OperatorConfigMapReconciler) deletionPhase() error {
```

Comment:
we need to proceed only if no PVs exist in the cluster using our provisioner name